### PR TITLE
New attribute to start fw broadcast process

### PR DESCRIPTION
**Pull request by SidharajU**
We need additional attribute to support 5-step FW broadcast process. This attribute is only applicable to PHY and its valid when broadcast is enabled. This is mainly used when the design has multiple PHYs and user would like to take advantage of broadcast FW download to save MDIO cycles and time.

### CODE REVIEW

**Comment by SidharajU** [comment]
I followed all the instructions available and its failing saying permission denied. Do I need to get any other permission to resolve this error?

bash-4.1$ gits
On branch patch-2
nothing to commit, working tree clean
bash-4.1$ **git commit --amend --no-edit --signoff**
[patch-2 7563b78] New attribute to start fw broadcast process
 Author: SidharajU <57233013+SidharajU@users.noreply.github.com>
 Date: Tue May 4 11:37:20 2021 -0700
 1 file changed, 12 insertions(+)
bash-4.1$ **git push --force-with-lease origin patch-2**
Warning: Permanently added the RSA host key for IP address 'a.b.c.d' to the list of known hosts.
**ERROR: Permission to opencomputeproject/SAI.git denied to SidharajU.
fatal: Could not read from remote repository.**

Please make sure you have the correct access rights
and the repository exists.
bash-4.1$

**Comment by lguohan** [comment]
can your provide call example?

how is the new attribute address the broadcast scenario?

**Comment by SidharajU** [comment]
Uploaded a document to list the steps needed for FW download in broadcast mode

**Comment by SidharajU** [comment]
> can your provide call example?
> 
> how is the new attribute address the broadcast scenario?

I have uploaded the document. Can you review it?

**Review by kcudnik — APPROVED** [review]

File: doc/macsec-gearbox/External PHY broadcast FW download steps [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+SAI_SWITCH_ATTR_FIRMWARE_DOWNLOAD_BROADCAST (Existing attribute)
+
+2. Broadcast enable step on each PHY
+Enable Broadcast on each PHY using the new proposed attribute SAI_SWITCH_ATTR_FIRMWARE_BROADCAST_BEGIN via set_switch_attribute call.
+
+3. Execute step only on the master PHY
+Call set_switch_attribute and set SAI_SWITCH_ATTR_FIRMWARE_DOWNLOAD_EXECUTE (Existing attribute)
+
+4. Verify FW step on each PHY
+Call set_switch_attribute and set SAI_SWITCH_ATTR_FIRMWARE_VERIFY_AND_INIT_SWITCH (Existing attribute)
```
Comment (lguohan) [inline doc/macsec-gearbox/External PHY broadcast FW download steps:27]: why step 4 is before step 5? we should stop the broadcast first and then verify?

Comment (SidharajU) [reply doc/macsec-gearbox/External PHY broadcast FW download steps:27]: So that we can verify the FW upgrade first before disabling broadcast.

**Comment by jleveque** [comment]
@SidharajU: Please review this proposal and provide feedback:

- Add a new attribute of type [`sai_s8_list_t`](https://github.com/opencomputeproject/SAI/blob/master/inc/saiswitch.h#L1681) where the vendor can specify all the MDIO slave addresses for a given PHY

Then the following would be the FW update logic:
1. For each MDIO master, call `create_switch()` and pass the list of slave MDIO addresses via the attribute detailed above
    - In this function, the 5-step FW upgrade procedure as you detail in your document will be carried out for **all** PHYs on this MDIO bus.
 2. Call `create_switch()` on all slave PHYs, which will already have had their firmware upgraded.

**Comment by rlhui** [comment]
@jleveque , @lguohan  - please help review/approve, thanks.

**Comment by jleveque** [comment]
@rlhui: There has been no activity since my last comment.

**Comment by jleveque** [comment]
@@SidharajU: Thanks for adding the new attribute. If you agree with the FW update logic as described in my comment, please update the document to describe this flow.

**Comment by SidharajU** [comment]
> @@SidharajU: Thanks for adding the new attribute. If you agree with the FW update logic as described in my comment, please update the document to describe this flow.

Hi Joe, 
I have removed the document as its no longer applicable and needed. The attribute is self-explanatory so no document is added.

I am getting DCO error. I tried the following steps but its still failing
git fetch origin pull/1232/head:patch-2
git checkout patch-2
git commit --amend --no-edit --signoff
git push -f sid patch-2

Any idea what's wrong with DCO error?

**Comment by jleveque** [comment]
> > @@SidharajU: Thanks for adding the new attribute. If you agree with the FW update logic as described in my comment, please update the document to describe this flow.
> 
> Hi Joe,
> I have removed the document as its no longer applicable and needed. The attribute is self-explanatory so no document is added.

OK, as long as you think it's self-explanatory.

> 
> I am getting DCO error. I tried the following steps but its still failing
> git fetch origin pull/1232/head:patch-2
> git checkout patch-2
> git commit --amend --no-edit --signoff
> git push -f sid patch-2
> 
> Any idea what's wrong with DCO error?

I do not know what could be causing this issue. Maybe @kcudnik has some advice?

**Comment by kcudnik** [comment]
have you tried commands from DCO suggestions: https://github.com/opencomputeproject/SAI/pull/1232/checks?check_run_id=2956038705 ?

**Comment by SidharajU** [comment]
> have you tried commands from DCO suggestions: https://github.com/opencomputeproject/SAI/pull/1232/checks?check_run_id=2956038705 ?

Yes. I tried the same steps multiple times but its still failing.

**Comment by kcudnik** [comment]
no clue then, everything seems fine, i can mark it manually to pass

**Comment by SidharajU** [comment]
@rlhui, @jleveque, can you approve and merge the PR?

**Review by jleveque — APPROVED** [review]

**Comment by rlhui** [comment]
> @SidharajU: Please review this proposal and provide feedback:
> 
> * Add a new attribute of type [`sai_s8_list_t`](https://github.com/opencomputeproject/SAI/blob/master/inc/saiswitch.h#L1681) where the vendor can specify all the MDIO slave addresses for a given PHY
> 
> Then the following would be the FW update logic:
> 
> 1. For each MDIO master, call `create_switch()` and pass the list of slave MDIO addresses via the attribute detailed above
>    
>    * In this function, the 5-step FW upgrade procedure as you detail in your document will be carried out for **all** PHYs on this MDIO bus.
> 2. Call `create_switch()` on all slave PHYs, which will already have had their firmware upgraded.

@SidharajU  - now the doc is removed.  Are the above steps the recommended steps?  I don't see the sequence is documented anywhere, in comment or other?  Thanks.

**Comment by SidharajU** [comment]
@kcudnik I am seeing the same DCO error again. Is it because I am using the old authentication method(received deprecation notice in the email)? Could you pls manually approve this?

**Comment by SidharajU** [comment]
I ran the commands mentioned in the details but the DCO is still failing:

bash-4.1$ **git commit --amend --no-edit --signoff**

[patch-2 624f41e] Updating FW broadcast sequence
 Author: SidharajU <57233013+SidharajU@users.noreply.github.com>
 Date: Wed Jun 30 12:36:00 2021 -0700
 2 files changed, 6 insertions(+), 31 deletions(-)
 delete mode 100644 doc/macsec-gearbox/External PHY broadcast FW download steps
bash-4.1$ 
bash-4.1$ **git push -f sid patch-2**
Enumerating objects: 11, done.
Counting objects: 100% (11/11), done.
Delta compression using up to 2 threads.
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 826 bytes | 826.00 KiB/s, done.
Total 6 (delta 5), reused 0 (delta 0)
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
To https://github.com/SidharajU/SAI.git
 + ac6cc64...624f41e patch-2 -> patch-2 (forced update)
bash-4.1$

**Comment by kcudnik** [comment]
i think this could be because of this commit: https://github.com/opencomputeproject/SAI/pull/1232/commits/d9cf2b7939b983467294c83e228ddc3363fb5782 maybe DCO requires all commits to be signed off

**Comment by SidharajU** [comment]
> i think this could be because of this commit: [d9cf2b7](https://github.com/opencomputeproject/SAI/commit/d9cf2b7939b983467294c83e228ddc3363fb5782) maybe DCO requires all commits to be signed off

I always use --signoff. Not sure what/why the error is. How do we fix this issue?

**Comment by kcudnik** [comment]
the commit i mentioned don't have your sign off signature, but im not sure whether this is DCO issue or not :(

**Comment by SidharajU** [comment]
> the commit i mentioned don't have your sign off signature, but im not sure whether this is DCO issue or not :(

Then how do I fix this issue? Can you manually pass it since we don't know what's the issue is?

**Comment by SidharajU** [comment]
@kcudnik @rlhui  Any update? Who can help fix this unknown issue? Or we an manually approve this error?

**Comment by kcudnik** [comment]
its all signed off now

**Pull request merged**